src/lib/operators/table_scan/column_between_table_scan_impl.hpp

```diff
@@ -32,10 +32,12 @@ class ColumnBetweenTableScanImpl : public AbstractSingleColumnTableScanImpl {
   void _scan_non_reference_segment(const BaseSegment& segment, const ChunkID chunk_id, PosList& matches,
                                    const std::shared_ptr<const PosList>& position_filter) const override;
 
-  void _scan_segment(const BaseSegment& segment, const ChunkID chunk_id, PosList& matches,
-                     const std::shared_ptr<const PosList>& position_filter) const;
-  void _scan_segment(const BaseDictionarySegment& segment, const ChunkID chunk_id, PosList& matches,
-                     const std::shared_ptr<const PosList>& position_filter) const;
+  void _scan_non_dictionary_segment(const BaseSegment& segment, const ChunkID chunk_id, PosList& matches,
+                                    const std::shared_ptr<const PosList>& position_filter) const;
+
+  // Optimized scan on DictionarySegments
+  void _scan_dictionary_segment(const BaseDictionarySegment& segment, const ChunkID chunk_id, PosList& matches,
```

Comment:
Argh, I was happy to find a way where we do not have to add types to the method name and where the fallback was implicit, not explicit. What is the advantage of this?

Reply:
> First, the previous version instantiated the lambda for each segment type, only to distinguish between dictionary and non-dictionary segments. Minor, but this is why I first changed it.

Did this show up in the compile time and if so, how much was the difference?

At least you convinced me to think about this a bit more. Generally, I am not a fan of the `if (dynamic_cast...` pattern

Reply:
Should we decide that this is the way to go, I would whine about the use of raw pointers next. Is there any need to convert the shared_ptr to a raw pointer to a reference?

Reply:
> Did this show up in the compile time and if so, how much was the difference?

Nah, this was just me purging "resolve_*" stuff from the code base, I don't think this particular case would have had a negative impact, its just ~7 instantiations of a lambda

> Generally, I am not a fan of the if (dynamic_cast... pattern

Neither am I, but I think in this case this is better than using resolve_segment_type, just to distinguish between two types of segments


> Should we decide that this is the way to go, I would whine about the use of raw pointers next. Is there any need to convert the shared_ptr to a raw pointer to a reference?

This doesn't really bother me. We could hand the shared_ptr all the way through the TableScanImpl, but I reckon that's more typing in the end.

Reply:
Probably misunderstand you, but "dynamic_pointer_cast" won't work on references, right? So if we stick with passing "BaseSegment" around as reference (which I am in favour of), then we have no choice but to use "dynamic_cast<x*>", since "dynamic_cast<x&>" reports failures with an exception.